File: solutions/35-search-insert-problem/search-insert-problem.go
package main

func main() {
	searchInsert()
}

func searchInsert(nums []int, target int) int {
	return binaryInsert(nums, target, 0, len(nums)-1)
}

func binaryInsert(nums []int, target int, lo int, hi int) int {
	if lo > hi {
		return lo
	}

	mid := (lo + hi) / 2
	if target == nums[mid] {
		return mid
	}
	if target < nums[mid] {
		return binaryInsert(nums, target, lo, mid-1)
	}

	return binaryInsert(nums, target, mid+1, hi)
}

File: solutions/35-search-insert-problem/search-insert-problem.exs
defmodule Solution do
  @spec search_insert(nums :: [integer], target :: integer) :: integer
  def search_insert(nums, target) do
    binary_insert(nums, target, 0, length(nums) - 1)
  end

  defp binary_insert(_nums, _target, lo, hi) when lo > hi, do: -1

  defp binary_insert(nums, target, lo, hi) do
    mid = div(lo + hi, 2)
    val = Enum.at(nums, mid)

    cond do
      target > val -> binary_insert(nums, target, mid + 1, hi)
      target < val -> binary_insert(nums, target, lo, mid - 1)
      true -> mid
    end
  end
end
